### PR TITLE
cli: support more user friendly userfile upload semantics

### PR DESCRIPTION
**Pull request by adityamaru**
Previously, the user would have to specify both the source and
detination args when using `userfile upload`.  This change tweaks the
semantics to be as follows:

- Users must specify a source path.

- If the user does not specify a destination URI/path, we use the
  default URI scheme and host, and the basename from the source arg as
the path.

- If the destination is a well-formed userfile URI of the form
  userfile://db.schema.tablename_prefix/path/to/file, then we use that
as the final URI.

- If destination is not a well-formed userfile URI, we use the default
  userfile URI schema and host, and the destination as the path.

File paths are never cleaned during upload or interaction with the
UserFileStorage. Thus, the filename is the exact string represented by
the path of the constructed userfile URI. This is in keeping with the
semantics of s3 and gcsutil. To prevent user surprises we still reject
constructs where the URI post normalization != URI pre normalization.

egs: test/./test.csv or test/../../test.csv

Informs: #51222

Release note (cli change): Improves the user semantics for `userfile
upload` by supporting different patterns of specifying the source and
destination CLI args. The source arg is required, while the destination
arg is now optional.

### CODE REVIEW

**Comment by adityamaru** [comment]
Commit depends on #50981, I will rebase once merged.

**Review by miretskiy — CHANGES_REQUESTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-:-MC7Qiew1_m220u-y53D:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[pkg/cli/userfiletable_test.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MC7QdJr8B1nXlTRVcrK:-MC7QdJr8B1nXlTRVcrL:b-5tuws9) ([raw file](https://github.com/cockroachdb/cockroach/blob/925ff780d26aaaba5728392e1c46f396729242a6/pkg/cli/userfiletable_test.go#L41)):*
> ```Go
> 	// User passes no destination so we use the basename of the source as the
> 	// filename.
> 	c.Run(fmt.Sprintf("userfile upload %s", file))
> ```

any way to check if the upload path is what you expect it to be?

File: pkg/cli/flags.go
Comment (dt) [inline pkg/cli/flags.go:544]: i think you somehow doubled this line in a rebase since you also have an extra commit

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @miretskiy)

___
*[pkg/cli/flags.go, line 544 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MC3TuTZiEnKTt7eNQSK-r3-544:-MCAB5sI3tOIqQvJYWMQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/925ff780d26aaaba5728392e1c46f396729242a6/pkg/cli/flags.go#L544)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

i think you somehow doubled this line in a rebase since you also have an extra commit
</blockquote></details>

Done.
___
*[pkg/cli/userfiletable_test.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MC7QdJr8B1nXlTRVcrK:-MCAB77B4i0KJ1ARFDTW:bydbpai) ([raw file](https://github.com/cockroachdb/cockroach/blob/925ff780d26aaaba5728392e1c46f396729242a6/pkg/cli/userfiletable_test.go#L41)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

any way to check if the upload path is what you expect it to be?
</blockquote></details>

I'm not sure I understand, the comments at the end of this test are parsed as the expected outputs.

**Review by knz — COMMENTED** [review]
@aaron-crl I'd like to put this in your radar as well. Ideally, you'd come up with some guidelines / evaluation criteria to help Aditya assert on his own whether the path handling is sane.

> Reviewed 3 of 3 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-:-MCBR6ukDrYA-vRqWys9:b-xxe0dn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @dt, and @miretskiy)

___
*[pkg/cli/userfile.go, line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCBQtU22TFvBL-ZHI-S:-MCBQtU3BDC5x_ruh5jx:b7c988n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b262cb305a3a967e946ccb42864eb2fbe64694e/pkg/cli/userfile.go#L91)):*
> ```Go
> 		userFileURL := url.URL{
> 			Scheme: defaultUserfileScheme,
> 			Host:   defaultQualifiedNamePrefix + user,
> ```

see my comment about the username in the other PR.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @knz, and @miretskiy)

___
*[pkg/cli/userfile.go, line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCBQtU22TFvBL-ZHI-S:-MCNls612B3ySn2dOKq9:b-swd19r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b262cb305a3a967e946ccb42864eb2fbe64694e/pkg/cli/userfile.go#L91)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

see my comment about the username in the other PR.
</blockquote></details>

addressed in the previous PR by unescaping the URI before we use it. This allows for arbitrary Unicode characters.

**Comment by adityamaru** [comment]
RFAL, please

**Review by knz — COMMENTED** [review]
> Reviewed 8 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @dt, @knz, and @miretskiy)

___
*[pkg/storage/cloudimpl/file_table_storage.go, line 118 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCQe-EMDl-7tGx-X1HE:-MCQe-EMDl-7tGx-X1HF:bczg2wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1990865dfffbc6e3e3be12eb2362da33463080ba/pkg/storage/cloudimpl/file_table_storage.go#L118)):*
> ```Go
> 
> 	if path.Clean(basename) != basename {
> 		return "", errors.Newf("basename %s changes to %s on normalization. "+
> ```

I'm OK with this test but I'd like a unit test that demonstrates the error: under which circumstances would client code encounter this?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @knz, and @miretskiy)

___
*[pkg/storage/cloudimpl/file_table_storage.go, line 118 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCQe-EMDl-7tGx-X1HE:-MCRvEMz4QuL7DJcRsXE:bcmmtil) ([raw file](https://github.com/cockroachdb/cockroach/blob/1990865dfffbc6e3e3be12eb2362da33463080ba/pkg/storage/cloudimpl/file_table_storage.go#L118)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm OK with this test but I'd like a unit test that demonstrates the error: under which circumstances would client code encounter this?
</blockquote></details>

CLI related code should not encounter this because all the FileTableStorage methods are passed an empty basename, as a result of which all operations occur on the prefix.
BACKUP uses the basename when writing data, manifest files, and encryption information. I have added a specialized test to `file_table_storage_test.go` to ensure we're rejecting what we think we are. `external_storage_test.go` has tests in `testExportStore` which test valid basename constructs.

**Review by knz — COMMENTED** [review]
> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @dt, @knz, and @miretskiy)

___
*[pkg/cli/userfiletable_test.go, line 159 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCkbuXy1KShK9tiK4NQ:-MCkbuXzCqJaXMtNcjrT:bmsjdmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e90fe8517e70dc556497dd789d91ab8287513b6/pkg/cli/userfiletable_test.go#L159)):*
> ```Go
> 	}
> 
> 	require.NoError(t, os.RemoveAll(dir))
> ```

Can you double-check what the `cleanFn` returned by `TempDir` does? It seems to me that you do not need this RemoveAll here.
___
*[pkg/storage/cloudimpl/file_table_storage.go, line 118 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCQe-EMDl-7tGx-X1HE:-MCkc2HbC-NY6q-470Nt:b-mj5ddq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1990865dfffbc6e3e3be12eb2362da33463080ba/pkg/storage/cloudimpl/file_table_storage.go#L118)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

Client code should not encounter this because all the FileTableStorage methods are passed an empty basename, as a result of which all operations occur on the prefix.
BACKUP uses the basename when writing data, manifest files, and encryption information. I have added a specialized test to `file_table_storage_test.go` to ensure we're rejecting what we think we are. `external_storage_test.go` has tests in `testExportStore` which test valid basename constructs.  
</blockquote></details>

Yes that works thanks

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @knz, and @miretskiy)

___
*[pkg/cli/userfiletable_test.go, line 159 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCkbuXy1KShK9tiK4NQ:-MClSOkOB5AlQt97WphV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e90fe8517e70dc556497dd789d91ab8287513b6/pkg/cli/userfiletable_test.go#L159)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you double-check what the `cleanFn` returned by `TempDir` does? It seems to me that you do not need this RemoveAll here.
</blockquote></details>

Done.

**Review by knz — COMMENTED** [review]
Thanks. You'll want your final PR approval from someone in your team.

> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-:-MCqQCaoBQwulK1ziQVp:b-kskzif)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @knz, and @miretskiy)

**Review by miretskiy — APPROVED** [review]
> Reviewed 1 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51353)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @dt, and @knz)

___
*[pkg/cli/userfiletable_test.go, line 131 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCqc_8VCxJML9t1U05g:-MCqc_8VCxJML9t1U05h:b-tq19pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea15683f47371941cb58e9d0ac83e608a40c67b1/pkg/cli/userfiletable_test.go#L131)):*
> ```Go
> 		if err != nil {
> 			t.Fatal(err)
> 		}
> ```

require.NoError(t, err) here and throughout?
___
*[pkg/storage/cloudimpl/cloudimpltests/file_table_storage_test.go, line 64 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51353#-MCqcwaRDIhk3CiCrgE6:-MCqcwaRDIhk3CiCrgE7:b-j9jr8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea15683f47371941cb58e9d0ac83e608a40c67b1/pkg/storage/cloudimpl/cloudimpltests/file_table_storage_test.go#L64)):*
> ```Go
> 		require.True(t, testutils.IsError(store.WriteFile(ctx, testfile, bytes.NewReader([]byte{0})),
> 			"basename listing-test/../basepath changes to basepath on normalization. "+
> 				"userfile does not permit such constructs."))
> ```

you don't have to specify full error message.  Just substring (does not permit such).
Also, I would pull out write file call and pass in err to require.True so that if the test fails, you'll see what kind of error you got (as opposed to "expected true")
  err := store.WriteFile()
 require.True(t, testutils.IsError(err, "changes to basepath", err))

**Comment by adityamaru** [comment]
TFTRs!

bors r+

**Pull request merged**
